File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/AuthorizationPreservationProof.java
package com.foundgine.core.semantic.planning;

import java.util.*;

/** Records that a rewrite did not weaken the source plan's security contract. */
public record AuthorizationPreservationProof(boolean isSatisfied, List<String> violations) {
    public AuthorizationPreservationProof { violations=violations==null?List.of():List.copyOf(violations); }
    public static AuthorizationPreservationProof create(SemanticPlan before,SemanticPlan after){
        Objects.requireNonNull(before,"before");Objects.requireNonNull(after,"after");
        var afterInvariants=new HashSet<>(after.effectiveSecurityInvariants());var dropped=before.effectiveSecurityInvariants().stream().filter(x->!afterInvariants.contains(x)).toList();
        if(dropped.isEmpty())return new AuthorizationPreservationProof(true,List.of());
        return new AuthorizationPreservationProof(false,dropped.stream().map(id->"security invariant '"+id+"' required by the source plan is not preserved by the rewritten plan.").toList());
    }
}

File: src/csharp/Foundgine.Core/Semantic/Planning/AuthorizationPreservationProof.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Records that a rewrite did not weaken the source plan's security contract: every security
/// invariant required by the plan before rewriting is still required after. A rewrite is free
/// to add invariants (e.g. because it introduces a new authorization-sensitive shape) but must
/// never silently drop one — dropping an invariant is exactly how a "purely semantic"
/// optimization could quietly regress a security guarantee.
/// </summary>
public sealed record AuthorizationPreservationProof(bool IsSatisfied, IReadOnlyList<string> Violations)
{
    public static AuthorizationPreservationProof Create(SemanticPlan before, SemanticPlan after)
    {
        ArgumentNullException.ThrowIfNull(before);
        ArgumentNullException.ThrowIfNull(after);

        var afterInvariants = new HashSet<string>(after.EffectiveSecurityInvariants, StringComparer.Ordinal);

        var dropped = before.EffectiveSecurityInvariants
            .Where(id => !afterInvariants.Contains(id))
            .ToArray();

        if (dropped.Length == 0)
            return new AuthorizationPreservationProof(true, []);

        return new AuthorizationPreservationProof(
            false,
            dropped
                .Select(id =>
                    $"security invariant '{id}' required by the source plan is not preserved by the rewritten plan.")
                .ToArray());
    }
}
